core/types/rollup_l1_cost.go

```diff
@@ -53,31 +59,55 @@ var (
-	cacheBlockNum := ^uint64(0)
-	var l1BaseFee, overhead, scalar *big.Int
-	return func(blockNum uint64, blockTime uint64, dataGas RollupGasData, isDepositTx bool) *big.Int {
-		rollupDataGas := dataGas.DataGas(blockTime, config) // Only fake txs for RPC view-calls are 0.
-		if config.Optimism == nil || isDepositTx || rollupDataGas == 0 {
-			return nil
+// NewL1CostFunc returns a function used for calculating L1 fee cost.  This depends on the oracles
+// because gas costs can change over time, and depends on blockTime since the specific function
+// used to compute the fee can differ between hardforks.
+func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter, blockTime uint64) L1CostFunc {
+	l1BaseFee := statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
+	overhead := statedb.GetState(L1BlockAddr, OverheadSlot).Big()
+	scalar := statedb.GetState(L1BlockAddr, ScalarSlot).Big()
+	f := newL1CostFunc(config, l1BaseFee, overhead, scalar, blockTime)
+	return func(l1CostData L1CostData) *big.Int {
+		fee, _ := f(l1CostData)
+		return fee
+	}
+}
+
+func newL1CostFunc(config *params.ChainConfig, l1BaseFee, overhead, scalar *big.Int, blockTime uint64) l1CostFunc {
+	isRegolith := config.IsRegolith(blockTime)
+	return func(l1CostData L1CostData) (fee, gasUsed *big.Int) {
+		if config.Optimism == nil {
+			return nil, nil
 		}
-		if blockNum != cacheBlockNum {
-			l1BaseFee = statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
-			overhead = statedb.GetState(L1BlockAddr, OverheadSlot).Big()
-			scalar = statedb.GetState(L1BlockAddr, ScalarSlot).Big()
-			cacheBlockNum = blockNum
+		gas := l1CostData.zeroes * params.TxDataZeroGas
+		if isRegolith {
+			gas += l1CostData.ones * params.TxDataNonZeroGasEIP2028
+		} else {
+			gas += (l1CostData.ones + 68) * params.TxDataNonZeroGasEIP2028
 		}
-		return L1Cost(rollupDataGas, l1BaseFee, overhead, scalar)
+		l1GasUsed := new(big.Int).SetUint64(gas)
+		l1GasUsed = l1GasUsed.Add(l1GasUsed, overhead)
+		l1Cost := new(big.Int).Set(l1GasUsed)
+		l1Cost.Mul(l1GasUsed, l1BaseFee).Mul(l1Cost, scalar).Div(l1Cost, big.NewInt(1_000_000))
+		return l1Cost, l1GasUsed
 	}
 }
 
-func L1Cost(rollupDataGas uint64, l1BaseFee, overhead, scalar *big.Int) *big.Int {
-	l1GasUsed := new(big.Int).SetUint64(rollupDataGas)
-	l1GasUsed = l1GasUsed.Add(l1GasUsed, overhead)
-	l1Cost := l1GasUsed.Mul(l1GasUsed, l1BaseFee)
-	l1Cost = l1Cost.Mul(l1Cost, scalar)
-	return l1Cost.Div(l1Cost, big.NewInt(1_000_000))
+// extractL1GasParams extracts the gas parameters necessary to compute gas costs from pre-Ecotone
+// L1 block info calldata.
+func extractL1GasParams(config *params.ChainConfig, time uint64, data []byte) (l1Basefee *big.Int, costFunc l1CostFunc, feeScalar *big.Float, err error) {
+	// pre-Ecotone, data consists of func selector followed by 7 ABI-encoded parameters (32 bytes each)
+	if len(data) < 4+32*8 {
+		// q from Roberto: shouldn't there be exactly 4+32*7 bytes? why do we allow more here?
```

Comment:
We should address this before merge

Reply:
yeah this is mostly an open question for you all... I was hesitant to change this because I was worried it might break existing chains, but it seems like this should have been ==.

Reply:
In the unit test I just added I made it codify the existing behavior which I think would be unsafe to change even if it was what we should have done.